keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+capacity.
+
+To simplify the model, make it backwards compatible and to avoid that it can evolve into something
+different and collide with other APIs, like Gateway APIs, we are adding the following constraints:
+
+- a ServiceCIDRConfig will be immutable after creation (to be revisited before Beta).
+- a ServiceCIDRConfig can only be deleted if there are no Service IP associated to it (enforced by finalizer).
+- there can be overlapping ServiceCIDRConfigs.
+- the apiserver will periodically ensure that a "default" ServiceCIDRConfig exists to cover the service CIDR flags
+  and the "kubernetes.default" Service.
+- any IPAddress existing in a cluster has to belong to a ServiceCIDRConfig.
+- any Service with a ClusterIP assigned is expected to have always a IPAddress object associated.
+- a ServiceCIDRConfig which is being deleted can not be used to allocate new IPs
+
+ This creates a 1-to-1 relation between Service and IPAddress, and a 1-to-N relation between
+ ServiceCIDRConfig and IPAddress.
+
+ The new allocator logic can be used by other APIs, like Gateway API.
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes user I want to be able to dynamically increase the number of IPs available for
+Services.
+
+#### Story 2
+
+As a Kubernetes admin I want to have a process that allows me to renumber my Services IPs.
+
+#### Story 3
+
+As a Kubernetes developer I want to be able to know the current Service IP range and the IP
+addresses allocated.
+
+#### Story 4
+
+As a Kubernetes admin that wants to use IPv6, I want to be able to follow the IPv6 address planning
+best practices, being able to assign /64 prefixes to my end subnets.
+
+### Notes/Constraints/Caveats (Optional)
+
+Current API machinery doesn't consider transactions, Services API is kind of special in this regard
+since already performs allocations inside the Registry pipeline using the Storage to keep
+consistency on an Obscure API Object that stores the bitmap with the allocations. This proposal still
+maintains this behavior, but instead of modifying the shared bitmap, it will create a new IPAddress object.
+
+Changing service IP allocation to be async with regards to the service creation would be a MAJOR
+semantic change and would almost certainly break clients.
+
+### Risks and Mitigations
+
+Cross validation of Objects is not common in the Kubernetes API, sig-apimachinery should verify that a
+bad precedent is introduced.
+
+Current Kubernetes cluster have a very static network configuration, allowing to expand the Service
+IP ranges will give more flexibility to users, with the risk of having problematic or incongruent
+network configurations with overlapping. But this is not really a new problem, users need to do a
+proper network planning before adding new ranges to the Service IP pool.
+
+Service implementations,like kube-proxy, can be impacted if they were doing assumptions about the ranges assigned to the Services.
```

Comment:
DNS implementations also need to know the service CIDRs for PTR lookups (and to know which PTR look ups they are NOT authoritative on). So those need to be updated as well.

Reply:
I can't find any reference to Service CIDR on the DNS spec or in CoreDNS.
It seems that the authoritative part here is the `ClusterIP` associated to the `Service`, and this is provided by the apiserver allocators, that guarantee that the IPAddress belongs to a configured ServiceCIDRConfig

Reply:
oh, how do those work today?  Do we configure them with the service CIDR?  I just assumed it would reply PTR for any service it had an A/AAAA for

Reply:
It's not that simple. It needs to know which subnets it is authoritative for, so that it can send other PTR requests to an upstream server, and return NXDOMAIN for non-existant PTRs. See for example:

https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed#L62

`CLUSTER_DOMAIN` and `REVERSE_CIDRS` represent the domains for which it is authoritative.

Ideally, we list all CIDRs for which we serve PTRs in there. In reality, people don't always configure those, they may put something broader. Also, because DNS is not super great with classless subnets (no well followed standard), I believe the default logic will "round" a collection of CIDRs to the classful CIDR that contain them. In other words, CoreDNS may have a larger ARPA domain configured than what is actually validly in its authority.

This means we may mask upstream PTRs. To work around this, we add `fallthrough` which basically says "oh, I would return NXDOMAIN, let me pass this upstream and return whatever they give me instead". This is not ideal but is sort of the best we can do. It does mean that even if we *are* authoritative (the PTR request falls in the smaller, non-classful CIDR range), we still pass it upstream. Which can lead to some anomolies.

I suppose we could fix this with a k8s plugin specific configuration of the CIDRs; it would let us accurately decide when to go upstream. But without that info available via the API, the config burden has been way too high to make it worthwhile. With change, we're getting closer!

Anyway...all I am asking for here is one line saying that DNS reverse lookup implementations may be impacted as well. :)

Reply:
oh, TIL, I didn't know that, I was checking the folllowin values 
```
        kubernetes cluster.local in-addr.arpa ip6.arpa {              
           pods insecure                                                        
           fallthrough in-addr.arpa ip6.arpa                                    
           ttl 30                                                               
        }        
```

I'll add a comment

Reply:
yeah, that works well enough in most cases. but it means that for any IP we don't have a PTR for, we will ask the upstream if they have one. Which can pollute your responses with upstream garbage or cause other issues. If we can narrow down from "in-addr.arpa", k8s plugin won't even try to answer the query (for example, on the line with `kubernetes` you had `10.0.0.0/8` instead of `in-addr.arpa` the k8s plugin would be skipped altogether for PTRs not in 10/8).